### PR TITLE
SI-8575 Fix subtyping transitivity with aliases, compound types

### PR DESCRIPTION
**Pull request by retronym**
Thanks to @b-studios and @Blaisorblade for investigation into this
bug.

`RefinedType#normalize` is responsible for flattening nested
compound types to a flat representation.

Types are normalized during `=:=` in search of a successful
result.

This means that `((A with B) with C) =:= (A with B with C)`.

However, if we introduce a type alias for `A with B` on the LHS,
the attempt at flattening is thwarted.

This commit dealiases parents of refined types during normalization.

Two tests are included. The first demonstrates a symptom of this
problem: failure to install a bridge method resulted in a linkage
error.

The second test uses the reflection API to directly show transitivity
of subtyping now holds.

Targetting at 2.12, as bug fixes in subtyping usually can be shown
to lead to binary incompatibilities between code using the old and
new compilers.

### CODE REVIEW

**Comment by Blaisorblade** [comment]
The failures don't look spurious. This is a failure of a pos test. `Box[this.l]` should be `Box[ ({ type l[A[x]] = K[ (A ∙ B)#l] })#l ]`, so is either pretty-printed "wrong"/confusing (and the error is elsewhere), or shows a bug in substitution, with closed terms becoming open. Without looking at code yet, I'd start checking that `dealias` on `({ type l[A[x]] = K[ (A ∙ B)#l] })#l` has the considerable smarts needed to figure out it probably can't do anything, unless you add type lambdas to the intermediate representation. (Don't both νObj and Dotty end up doing exactly that?)

> ```
> fail - pos/t7517.scala  [compilation failed]% scalac t7517.scala
> t7517.scala:16: error: polymorphic expression cannot be instantiated to expected type;
>  found   : [B[x]]Box.SplitBox[L,B]
>     (which expands to)  [B[x]]Box[this.l]
>  required: Box[Composed.this.Split]
>       val a: Box[Split] = Box.split(box)
>                                    ^
> t7517.scala:19: error: type mismatch;
>  found   : Box.SplitBox[L,B]
>     (which expands to)  Box[this.l]
>  required: Box[Composed.this.Split]
>       val a1: Box[Split] = Box.split[L,B](box)
>                                          ^
> t7517.scala:20: error: polymorphic expression cannot be instantiated to expected type;
>  found   : [B[x]]Box.SplitBox[L,B]
>     (which expands to)  [B[x]]Box[this.l]
>  required: Box[this.l]
>       val a2: Box[ ({ type l[A[x]] = L[ (A ∙ B)#l ] })#l ] = Box.split(box)
>                                                                       ^
> three errors found
> ```

Most other errors have to do with type lambdas.

Another failure is changing errors (Object vs AnyRef) — I imagine there `--update-check` is in order (this might be what the exception for `AnyRef` in the other patch was catering for).

**Comment by retronym** [comment]
I've just pushed another version which treads more lightly, by only dealiasing parents that point to refined types.

**Comment by Blaisorblade** [comment]
> I've just pushed another version which treads more lightly, by only dealiasing parents that point to refined types.

Could one make `dealias` less of a landmine for future hackers?

Also, in this solution, would a type lambda pointing to a refined type break there? Some variant of `{type l[X] = Foo with Bar}#l`? Maybe not right now, since it seems we aren't getting to those bodies in the first place.

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -1600,7 +1600,12 @@ trait Types
     private var normalized: Type = _
     private def normalizeImpl = {
       // TODO see comments around def intersectionType and def merge
-      def flatten(tps: List[Type]): List[Type] = tps flatMap { case RefinedType(parents, ds) if ds.isEmpty => flatten(parents) case tp => List(tp) }
+      // SI-8575 The dealias is needed here to keep subtyping transitive, example in run/t8575b.scala
+      def flatten(tps: List[Type]): List[Type] = {
+        def dealiasRefinement(tp: Type) = if (tp.dealias.isInstanceOf[RefinedType]) tp.dealias else tp
+        tps map dealiasRefinement flatMap {
+          case RefinedType(parents, ds) if ds.isEmpty => flatten(parents) case tp => List(tp) }
```
Comment (Blaisorblade) [inline src/reflect/scala/reflect/internal/Types.scala]: Is `case` really allowed without a newline? Not sure whether I should like it or not, but I think I don't...

Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala]: I've found an extra `\n` to donate to the patch :)

**Comment by Blaisorblade** [comment]
> Could one make dealias less of a landmine for future hackers?

I take the silence as a "no", and who am I to schedule your time? But I assigned you a new issue (https://issues.scala-lang.org/browse/SI-8850 — feel free to reassign after checking it's not stupid).

With that reservation, I'll throw in a

LGTM

for you, as a present :-)

**Comment by lrytz** [comment]
Would it be more correct to `dealiasRefinement` recursively? For things like `type A = X with Y; type B = A with Z`

**Comment by retronym** [comment]
I think we are okay piggy backing on the existing recursion of `flatten`. Perhaps you could extend the test to see if you can break it?

**Comment by lrytz** [comment]
I'll give it a try!

**Comment by retronym** [comment]
I've added the suggested test.

File: test/files/run/t8575c.scala [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+trait TypeMember {
+  type X
+  type Y
+  type Z
+}
+
+object Test extends TypeMember {
+  type A = X with Y
+  type B = Z with A
```
Comment (lrytz) [inline test/files/run/t8575c.scala:11]: it seems the typedefs are shadowing classes `A` and `B` – is that intended?

File: test/files/run/t8575b.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class A
+class B
+class C
+
+trait TypeMember {
+  type X
```
Comment (lrytz) [inline test/files/run/t8575b.scala]: `X` is unused

**Comment by gkossakowski** [comment]
Ping @retronym

**Comment by retronym** [comment]
pong. Will rework that test tomorrow.

**Comment by gkossakowski** [comment]
@retronym I hope for some :boom: once the test arrives :smirk:

**Comment by retronym** [comment]
Tests updated.

**Comment by retronym** [comment]
Re-review by @lrytz, please.

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by lrytz** [comment]
LGTM!

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
Rebased to try to avoid the hanging test suite.

**Pull request merged**